### DISCUSSION TITLE
MOSART Tagging Process

### DISCUSSION THREAD

**Discussion by ekluzek** [ESCOMP/MOSART#83] [Announcements]
Since, MOSART tags are managed by LMWG the process is largely the same as CTSM tags. For the "big picture" view for guidelines of bringing in advancements to MOSART see #84. Note that LMWG maintains: RTM, SLIM, mizuRoute and MOSART as externals (in addition to CTSM), so priorities of any of any of these take place in the context of LMWG larger priorities.

The checklist for CTSM tags is here:

https://github.com/ESCOMP/CTSM/blob/master/doc/README.CHECKLIST.master_tags

A simplified list for MOSART tags is below:

1. Checkout the latest CTSM tag on Derecho and Izumi (same machines as used for CTSM tags)
2. Run baselines with the "mosart" testlist to create_test on machines named in 1 above named (MOSART-tag)_(CTSM-tag) (for example mosart1_0_49_ctsm5.1.dev166) (run_sys_tests in CTSM can be used to do this)
3. In the CTSM sandbox checkout a branch in components/mosart and make your changes
4. Submit your mosart branch to github as a Pull Request (PR)
5. Get approval for the PR, and find out the ordering for the tag verses other tags and/or PR's (at least one MOSART maintainer should approve, additional reviewers are always encouraged as well) (prioritization of MOSART PR's is controlled by the LMWG see #84)
6. Run baselines of the update for the next tag as in step 2, but for the next MOSART tag when it's scheduled to come in

Steps 2 and 6 above can be done by maintainers, but collaborators could also do these steps.

Following steps are done by a MOSART maintainer (external collaborators can ask for permission to do this if they desire and agree to follow these guidelines):

8.  The maintainer updates the docs/ChangeLog file for the new tag
9.  The maintainer merges the PR to master in github (push the green merge button)
10. The maintainer compares the branch to master to show there are no diffs
11. The maintainer makes an annotated tag on master ) (i.e. git tag -a <tagname>)
12. The maintainer pushes that tag to ESCOMP
13. The maintainer updates the CTSM upcoming tags project, if necessary (https://github.com/ESCOMP/ctsm/projects/6)
14. The maintainer adds the tag to the CESM test database  (https://csegweb.cgd.ucar.edu/testdb)

Learning process for MOSART maintainers. We use the same learning process for CTSM tags:

1. The new maintainer watches while an experienced maintainer does a tag (steps 8 through the end)
2. The experienced maintainer watches which the new maintainer does a tag
3. Steps 1 and 2 are repeated as much as needed so the new maintainer is comfortable with the process

**Comment by ekluzek** [comment]
@mvertens @jedwards4b @slevis-lmwg @swensosc please give feedback on the above. Once, we think it's settled I'll remove the extraneous discussion and lock this discussion. We can still edit later, but we'll need to use other communication means to discuss it...

**Reply by slevis-lmwg** [reply]
@ekluzek let's please not lock it until after I've used the process once or twice, when I will have a good idea what I find unclear in the above. Thanks for putting this together!

**Reply by ekluzek** [reply]
Ahh, yes. Good point. We'll wait a while on it then...

**Comment by wwieder** [comment]
The only think I might add here is that MOSART, RTM, mizuRoute, SLIM and CTSM are all maintained by the CESM Land Model Working group.  As such, the prioritization of of MOSART PRs and attention from 'maintainers' is subject to the prioritization of broader priorities of the LMWG (that are ultimately managed by the LMWG co-chairs, SEs, and liaisons).

**Reply by ekluzek** [reply]
OK, I adjusted it a bit, linking to the other discussion and putting some of that text in both places. So I think I've addressed this, but feel free to let me know how it looks now.
